File: shared/sim/rules.js
// Pod + bonus simulation — shared by the SP client and the Go server.
// Pure: operates on plain state ({position, velocity, ...}). Spin / pulse /
// meshes are the render layer's job. RNG is Math.random for now; M2 swaps in a
// seeded xorshift so the server is deterministic.
import { Vector3 } from './vec.js';

// ---------------------------------------------------------------- pods -------

// pods: { list:[podState], centroid:Vector3, lost:number, total:number }
// podState: { position, velocity, spin, hp, radius, dead, captor }
// `spin` is a visual-tumble vector kept on the state so RNG order matches the
// original exactly; the server just doesn't stream it.
export function spawnPods(count, around, K) {
  const list = [];
  for (let i = 0; i < count; i++) {
    const spin = new Vector3(
      (Math.random() - 0.5) * K.spinRange,
      (Math.random() - 0.5) * K.spinRange,
      (Math.random() - 0.5) * K.spinRange,
    );
    list.push({
      position: new Vector3().randomDirection().multiplyScalar(K.spawnMin + Math.random() * K.spawnRange).add(around),
      velocity: new Vector3().randomDirection().multiplyScalar(K.driftMin + Math.random() * K.driftRange),
      spin,
      hp: K.hp,
      radius: K.radius,
      dead: false,
      captor: null,
    });
  }
  return list;
}

export function recentrePods(pods) {
  if (pods.list.length === 0) return;
  pods.centroid.set(0, 0, 0);
  for (const p of pods.list) pods.centroid.add(p.position);
  pods.centroid.multiplyScalar(1 / pods.list.length);
}

export function stepPods(pods, dt) {
  for (const p of pods.list) if (!p.dead && !p.captor) p.position.addScaledVector(p.velocity, dt);
  const keep = [];
  for (const p of pods.list) {
    if (p.dead) pods.lost++;
    else keep.push(p);
  }
  pods.list = keep;
  recentrePods(pods);
}

// ------------------------------------------------------------- bonuses ------

// bonuses: { list:[bonusState], timer, maxAlive }
// bonusState: { kind:'missiles'|'repair', position, velocity, radius, life, dead, _collected }
function spawnBonus(bonuses, around, player, K) {
  const wantRepair = player.hull < player.maxHull * K.wantRepairBelow;
  const wantMsl = player.missiles < player.maxMissiles * K.wantMissilesBelow;
  let kind;
  if (wantRepair && !wantMsl) kind = 'repair';
  else if (wantMsl && !wantRepair) kind = 'missiles';
  else kind = Math.random() < 0.5 ? 'repair' : 'missiles';

  bonuses.list.push({
    kind,
    _t: Math.random() * 6,        // visual phase (RNG order matches the original)
    position: new Vector3().randomDirection().multiplyScalar(K.spawnMin + Math.random() * K.spawnRange).add(around),
    velocity: new Vector3().randomDirection().multiplyScalar(K.driftMin + Math.random() * K.driftRange),
    radius: K.radius,
    life: K.life,
    dead: false,
    _collected: false,
  });
}

// a player projectile at `pos` scores a hit on a bonus -> counts as collecting
export function bonusHitByShot(bonuses, pos) {
  for (const b of bonuses.list) {
    if (b.dead) continue;
    if (pos.distanceToSquared(b.position) < b.radius * b.radius) { b._collected = true; return true; }
  }
  return false;
}

// returns { kind } for a bonus collected this step, or null
export function stepBonuses(bonuses, player, around, dt, K) {
  let collected = null;

  bonuses.timer -= dt;
  if (bonuses.timer <= 0 && bonuses.list.length < K.maxAlive) {
    bonuses.timer = K.respawnMin + Math.random() * K.respawnRange;
    spawnBonus(bonuses, around, player, K);
  }

  for (const b of bonuses.list) {
    if (b.dead) continue;
    b._t += dt;
    b.life -= dt;
    if (b.life <= 0) b.dead = true;
    b.position.addScaledVector(b.velocity, dt);
    if (b._collected || (player.alive && b.position.distanceToSquared(player.position) < b.radius * b.radius)) {
      b.dead = true;
      collected = { kind: b.kind };
    }
  }

  bonuses.list = bonuses.list.filter((b) => !b.dead);
  return collected;
}

export function makeBonuses(K) {
  return { list: [], timer: K.firstDelayMin + Math.random() * K.firstDelayRange, maxAlive: K.maxAlive };
}

// --------------------------------------------------------- level FSM -------

export function makeLevelFSM() {
  return { state: 'playing', timer: 0 };   // 'playing' | 'won' | 'lost'
}

// Call with dt = 0 to pause it (e.g. while the player is dead). ctx:
//   { podsAlive, enemiesCleared, level }
// Returns an action the caller performs, or null:
//   { flash: text, hold: seconds }   — show a banner
//   { startLevel: n }                — (re)start level n
export function stepLevelFSM(fsm, ctx, dt) {
  if (dt <= 0) return null;

  if (fsm.state === 'playing') {
    if (ctx.podsAlive === 0) {
      fsm.state = 'lost';
      fsm.timer = 3.0;
      return { flash: 'ALL PODS LOST — LEVEL FAILED', hold: 3.0 };
    }
    if (ctx.enemiesCleared) {
      fsm.state = 'won';
      fsm.timer = 2.8;
      return { flash: 'LEVEL ' + ctx.level + ' CLEARED', hold: 2.8 };
    }
    return null;
  }

  fsm.timer -= dt;
  if (fsm.timer <= 0) {
    const next = fsm.state === 'won' ? ctx.level + 1 : ctx.level;
    fsm.state = 'playing';
    return { startLevel: next };
  }
  return null;
}

File: server/game/rules.go
package game

import "strconv"

// Pods, bonuses and the level FSM — the Go twin of shared/sim/rules.js.
// RNG draw order matches the JS exactly (see comments).

// ---- pods -----------------------------------------------------------------

type Pod struct {
	Position Vec3
	Velocity Vec3
	Spin     Vec3
	HP       float64
	Radius   float64
	Dead     bool
	Captor   *Enemy
}

type Pods struct {
	List     []*Pod
	Centroid Vec3
	Lost     int
	Total    int
}

// spawnPods — 9 rng draws per pod: spin(3), dir(2)+mag(1), dir(2)+mag(1).
func spawnPods(count int, around Vec3, kp Block, rng *Rng) []*Pod {
	list := make([]*Pod, 0, count)
	for i := 0; i < count; i++ {
		spin := Vec3{
			(rng.Float64() - 0.5) * kp["spinRange"],
			(rng.Float64() - 0.5) * kp["spinRange"],
			(rng.Float64() - 0.5) * kp["spinRange"],
		}
		var pos, vel Vec3
		RandomDir(rng, &pos)
		pos.MultiplyScalar(kp["spawnMin"] + rng.Float64()*kp["spawnRange"]).Add(around)
		RandomDir(rng, &vel)
		vel.MultiplyScalar(kp["driftMin"] + rng.Float64()*kp["driftRange"])
		list = append(list, &Pod{
			Position: pos, Velocity: vel, Spin: spin,
			HP: kp["hp"], Radius: kp["radius"],
		})
	}
	return list
}

func recentrePods(p *Pods) {
	if len(p.List) == 0 {
		return
	}
	c := Vec3{}
	for _, pd := range p.List {
		c.Add(pd.Position)
	}
	c.MultiplyScalar(1 / float64(len(p.List)))
	p.Centroid = c
}

func stepPods(p *Pods, dt float64) {
	for _, pd := range p.List {
		if !pd.Dead && pd.Captor == nil {
			pd.Position.AddScaledVector(pd.Velocity, dt)
		}
	}
	keep := p.List[:0:0]
	for _, pd := range p.List {
		if pd.Dead {
			p.Lost++
		} else {
			keep = append(keep, pd)
		}
	}
	p.List = keep
	recentrePods(p)
}

// ---- bonuses ------------------------------------------------------------

type Bonus struct {
	Kind      string // "missiles" | "repair"
	T         float64
	Position  Vec3
	Velocity  Vec3
	Radius    float64
	Life      float64
	Dead      bool
	Collected bool
}

type Bonuses struct {
	List     []*Bonus
	Timer    float64
	MaxAlive int
}

func makeBonuses(kb Block, rng *Rng) *Bonuses {
	return &Bonuses{
		Timer:    kb["firstDelayMin"] + rng.Float64()*kb["firstDelayRange"],
		MaxAlive: int(kb["maxAlive"]),
	}
}

// spawnBonus draw order: kind (0 or 1 — only when both/neither wanted), _t(1),
// dir(2)+mag(1), dir(2)+mag(1).
func spawnBonus(b *Bonuses, around Vec3, ship *Ship, kb Block, rng *Rng) {
	wantRepair := ship.Hull < ship.MaxHull*kb["wantRepairBelow"]
	wantMsl := float64(ship.Missiles) < ship.MaxMissiles*kb["wantMissilesBelow"]
	var kind string
	switch {
	case wantRepair && !wantMsl:
		kind = "repair"
	case wantMsl && !wantRepair:
		kind = "missiles"
	default:
		if rng.Float64() < 0.5 {
			kind = "repair"
		} else {
			kind = "missiles"
		}
	}
	t := rng.Float64() * 6
	var pos, vel Vec3
	RandomDir(rng, &pos)
	pos.MultiplyScalar(kb["spawnMin"] + rng.Float64()*kb["spawnRange"]).Add(around)
	RandomDir(rng, &vel)
	vel.MultiplyScalar(kb["driftMin"] + rng.Float64()*kb["driftRange"])
	b.List = append(b.List, &Bonus{
		Kind: kind, T: t, Position: pos, Velocity: vel,
		Radius: kb["radius"], Life: kb["life"],
	})
}

func bonusHitByShot(b *Bonuses, pos Vec3) bool {
	for _, bo := range b.List {
		if bo.Dead {
			continue
		}
		if pos.DistanceToSq(bo.Position) < bo.Radius*bo.Radius {
			bo.Collected = true
			return true
		}
	}
	return false
}

// stepBonuses returns the kind collected this tick, or "".
func stepBonuses(b *Bonuses, ship *Ship, around Vec3, dt float64, kb Block, rng *Rng) string {
	collected := ""

	b.Timer -= dt
	if b.Timer <= 0 && len(b.List) < b.MaxAlive {
		b.Timer = kb["respawnMin"] + rng.Float64()*kb["respawnRange"]
		spawnBonus(b, around, ship, kb, rng)
	}

	for _, bo := range b.List {
		if bo.Dead {
			continue
		}
		bo.T += dt
		bo.Life -= dt
		if bo.Life <= 0 {
			bo.Dead = true
		}
		bo.Position.AddScaledVector(bo.Velocity, dt)
		if bo.Collected || (ship.Alive && bo.Position.DistanceToSq(ship.Pos) < bo.Radius*bo.Radius) {
			bo.Dead = true
			collected = bo.Kind
		}
	}

	keep := b.List[:0:0]
	for _, bo := range b.List {
		if !bo.Dead {
			keep = append(keep, bo)
		}
	}
	b.List = keep
	return collected
}

// ---- level FSM --------------------------------------------------------

type LevelFSM struct {
	State string // "playing" | "won" | "lost"
	Timer float64
}

func makeLevelFSM() *LevelFSM { return &LevelFSM{State: "playing"} }

// LevelAction is the FSM's output for a tick (zero value = nothing).
type LevelAction struct {
	Flash      string
	Hold       float64
	StartLevel int // 0 = none
}

func stepLevelFSM(f *LevelFSM, podsAlive int, enemiesCleared bool, level int, dt float64) (LevelAction, bool) {
	if dt <= 0 {
		return LevelAction{}, false
	}
	if f.State == "playing" {
		if podsAlive == 0 {
			f.State = "lost"
			f.Timer = 3.0
			return LevelAction{Flash: "ALL PODS LOST — LEVEL FAILED", Hold: 3.0}, true
		}
		if enemiesCleared {
			f.State = "won"
			f.Timer = 2.8
			return LevelAction{Flash: "LEVEL " + strconv.Itoa(level) + " CLEARED", Hold: 2.8}, true
		}
		return LevelAction{}, false
	}
	f.Timer -= dt
	if f.Timer <= 0 {
		next := level
		if f.State == "won" {
			next = level + 1
		}
		f.State = "playing"
		return LevelAction{StartLevel: next}, true
	}
	return LevelAction{}, false
}
